src/mapstorage/jsonmapstorage.cpp

```diff
@@ -0,0 +1,589 @@
+        {
+          Room copy = filter.alteredRoom( *rIter );
+          addRoom( jRooms, &copy );
+        }
+        else
+        {
+          assert( action == BaseMapSaveFilter::PASS );
+          addRoom( jRooms, *rIter );
+        }
+      }
+      else
+      {
+        addRoom( jRooms, *rIter );
+      }
+      progressCounter->step();
+    }
+
+    QString filePath = dir.filePath( zIter.key() + ".json" );
+    writeJson( filePath, jRooms, "zone" );
+  }
+}
+
+} // ns
+
+
+JsonMapStorage::JsonMapStorage(MapData& mapdata, const QString& filename) :
+    AbstractMapStorage(mapdata, filename)
+{}
+
+JsonMapStorage::~JsonMapStorage()
+{
+}
+
+void JsonMapStorage::newData()
+{
+  assert("JsonMapStorage does not implement newData()" != NULL);
+}
+
+bool JsonMapStorage::loadData()
+{
+  return false;
+}
+
+bool JsonMapStorage::mergeData()
+{
+  return false;
+}
+
+bool JsonMapStorage::saveData( bool baseMapOnly )
+{
+  emit log ("JsonMapStorage", "Writing data to files ...");
+
+
+  // Collect the room and marker lists. The room list can't be acquired
+  // directly apparently and we have to go through a RoomSaver which receives
+  // them from a sort of callback function.
+  // The RoomSaver acts as a lock on the rooms.
+  RoomList2 roomList;
+  MarkerList& markerList = m_mapData.getMarkersList();
+  RoomSaver saver(&m_mapData, roomList);
+  for (uint i = 0; i < m_mapData.getRoomsCount(); ++i)
+  {
+    m_mapData.lookingForRooms(&saver, i);
+  }
+
+  uint roomsCount = saver.getRoomsCount();
+  uint marksCount = markerList.size();
+  m_progressCounter->reset();
+  m_progressCounter->increaseTotalStepsBy( roomsCount * 2 + marksCount );
+
+  BaseMapSaveFilter filter;
+  if ( baseMapOnly )
+  {
+      filter.setMapData( &m_mapData );
+      m_progressCounter->increaseTotalStepsBy( filter.prepareCount() );
+      filter.prepare( m_progressCounter );
+  }
+
+  JsonWorld world;
+  world.addRooms( roomList, filter, m_progressCounter, baseMapOnly );
+
+  QDir saveDir( m_fileName );
+  QDir destDir( QFileInfo( saveDir, "v1" ).filePath() );
+  QDir roomIndexDir( QFileInfo( destDir, "roomindex" ).filePath() );
+  QDir zoneDir( QFileInfo( destDir, "zone" ).filePath() );
+  try {
+    if ( !saveDir.mkdir( "v1" ) )
+      throw std::runtime_error( "error creating dir v1" );
+    if ( !destDir.mkdir( "roomindex" ) )
+      throw std::runtime_error( "error creating dir v1/roomindex" );
+    if ( !destDir.mkdir( "zone" ) )
+      throw std::runtime_error( "error creating dir v1/zone" );
+
+    world.writeMetadata( QFileInfo( destDir, "arda.json" ), m_mapData );
+    world.writeRoomIndex( roomIndexDir );
+    world.writeZones( zoneDir, filter, m_progressCounter, baseMapOnly );
+  }
+  catch ( std::exception &e )
+  {
+    emit log( "JsonMapStorage", e.what() );
+    return false;
+  }
+
+  emit log ("JsonMapStorage", "Writing data finished.");
+
+  m_mapData.unsetDataChanged();
+  emit onDataSaved();
+
+  return true;
+}
+
+
+  /*
```

Comment:
Any reason to keep this comment?

Reply:
Looks like unfinished work. It definitively won't make it into v1 format, I'll remove it.